kernels/cuda/inference/hadamard_transform/hadamard_transform.cpp

```diff
@@ -0,0 +1,45 @@
+#include <torch/extension.h>
+#include <pybind11/pybind11.h>
+#include <ATen/cuda/CUDAContext.h>
+#include <c10/cuda/CUDAGuard.h>
+
+void run_fht(void* a, void* out, uint32_t numel, uint32_t had_size, cudaStream_t stream);
+
+torch::Tensor hadamard_transform(at::Tensor& in, bool inplace) {
+    auto dtype = in.scalar_type();
+    TORCH_CHECK(dtype == torch::ScalarType::Half, "Only fp16 supported currently");
+    TORCH_CHECK(in.is_cuda());
+    
+    const int had_size = in.size(-1);
+    TORCH_CHECK(had_size == 2 || had_size == 4 || had_size == 8 || had_size == 16
```

Comment:
minor, but consider a generic bitwise check for power of 2 and then check x < 2^15:
constexpr bool is_power_of_two(uint32_t x) {
    return x && !(x & (x - 1)); // Any power of 2 minus 1 is all ones: (2 N - 1 = 111....)
}
and x < (1U << 15).